### PR TITLE
Embedding Evaluation

### PR DESCRIPTION
**Pull request by tweddielin**
- `ProcessingPipeline` class is like `IterablePipeline` but a callable processing pipeline. Instead of  being lazy and building a generator as `IterablePipeline`, it eagerly returns the processed value right away. 

- `Clustering` class is a dict-like data structure. Key is the cluster concept and value is a list of entities associated to the concept. If `key_transform_fn` is provided, it will be applied to the key. If `value_transform_fn` is provided, it will be applied to the entities in the list for each concept. 

- Added 4 embedding metrics classes which take in a clustering object and the method `eval` returns the result dictionary where key is the concept and the value is the metric value.

-  `embedding_evaluate` function is to generate the total evaluation from the combination of metrics and clusterings.

- `convert_to_keyedvector ` is a function to produce a gensim's `KeyedVectors` object which is a structure defined in gensim for whatever embedding (could be document, graph node, ...)(https://radimrehurek.com/gensim/models/keyedvectors.html) so we can utilize the methods from the `KeyedVectors`. 

- Added 4 generic clusterings for O*Net as cached properties.

### CODE REVIEW

File: skills_ml/algorithms/preprocessing/__init__.py [hunk 1]
```diff
@@ -3,11 +3,55 @@
...
+
+     Attributes:
+        functions (generator): a series of functions
+
+    """
+    def __init__(self, *functions: Callable):
+        self.functions = functions
+
+    @property
+    def _compose(self):
```
Comment (thcrock) [inline skills_ml/algorithms/preprocessing/__init__.py]: I have never seen the property decorator applied to a pseudo-private function before, so it seems weird. The property decorator is generally aimed at improving the external interface of an object. 

But to be honest, if you just want compose functionality, I would either move that to its own function that can be imported, or even just use  https://github.com/pytoolz/toolz

So this property goes away entirely and you have.


```
from toolz import compose
...
def __call__(self, input_to_be_processed):
    return compose(self.functions)(input_to_be_processed)
```

File: skills_ml/evaluation/embedding_metrics.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+from collections import defaultdict
+from scipy.spatial import distance
+from gensim.models import KeyedVectors
+from typing import Callable, Dict
+import numpy as np
+import tempfile
+import os
+
+
+def embedding_evaluate(vectorization_pipeline, metric_objects):
```
Comment (thcrock) [inline skills_ml/evaluation/embedding_metrics.py]: I think we should name this to mirror the skill extractor evaluation routines: `metrics_for_embedding`

File: skills_ml/evaluation/embedding_metrics.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+class BaseEmbeddingMetric(object):
+    def __init__(self):
+        self.eval_result = None
+
+    def eval(self):
+        return NotImplementedError
+
+
+class CategorizationMetric(BaseEmbeddingMetric):
```
Comment (thcrock) [inline skills_ml/evaluation/embedding_metrics.py]: Docstrings for all of these! I honestly don't know what most of them are, so even I'm going to need the docstrings to understand them

File: skills_ml/evaluation/embedding_metrics.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    vector_size = vectorizing_pipeline.functions[-1].keywords['embedding_model'].vector_size
+    with tempfile.TemporaryDirectory() as td:
+        with open(os.path.join(td, 'w2v.format'), 'w') as f:
+            f.write(str(len(list_of_objects)) + " " + str(vector_size) + "\n")
+            for obj in list_of_objects:
+                f.write(identifier_fn(obj) + " ")
+                for s in list(map(str, list(vectorizing_pipeline(value_transform_fn(obj))))):
+                    f.write(s + " ")
+                f.write("\n")
+        wv = KeyedVectors.load_word2vec_format(os.path.join(td, 'w2v.format'))
```
Comment (thcrock) [inline skills_ml/evaluation/embedding_metrics.py]: I think you can just put the return here and Python will do the right thing with regards to closing the temporary directory.

Also I think you can just used NamedTemporaryFile for this and skip some of this code.

**Comment by thcrock** [comment]
Generally looks good but I did request some changes

File: skills_ml/algorithms/preprocessing/__init__.py (hunk 1 above)
Comment (tweddielin) [reply skills_ml/algorithms/preprocessing/__init__.py]: finally have a reason to use pytoolz! yeah vote for that.

**Comment by thcrock** [comment]
Looks like toolz needs to get added to the requirements

**Pull request merged**